src/charm.py

```diff
@@ -0,0 +1,323 @@
+                "POSTGRES_PORT": self._stored.db_port,
+                "POSTGRES_USER": self._stored.db_user,
+                "POSTGRES_PASSWORD": self._stored.db_password,
+                }
+        return pod_config
+
+    def _register_user(self, event):
+        """Registers a user for usage with Synapse."""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+
+        user = event.params["username"]
+        password = event.params["password"]
+        admin = event.params["admin"]
+        admin_switch = "--admin" if admin == "yes" else "--no-admin"
+
+        cmd = ["register_new_matrix_user",
+               "-u", user, "-p", password,
+               admin_switch, "-c", self._SYNAPSE_CONFIG_PATH,
+               "http://localhost:8008"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_generate_synapse(self):
+        """Runs the generate command for synapse"""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+
+        cmd = ["/start.py", "generate"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_migrate_synapse(self):
+        """Runs the migrate command for synapse"""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+        cmd = ["/start.py", "migrate_config"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
```

Comment:
This also seems to be done in a similar way a few times, potentially worth separating out into a function? Perhaps we can use/ write a shared library?

Reply:
I've separated it out, I like the idea of a shared library - I've seen such functionality in other charms we have and it makes sense to reuse it via a library.